File: backend/internal/models/roadmap.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type RoadmapSource string

const (
	RoadmapSourceLLMGenerated RoadmapSource = "llm_generated"
	RoadmapSourceCurated      RoadmapSource = "curated"
)

type RoadmapStatus string

const (
	RoadmapStatusActive    RoadmapStatus = "active"
	RoadmapStatusCompleted RoadmapStatus = "completed"
	RoadmapStatusAbandoned RoadmapStatus = "abandoned"
)

type NodeStatus string

const (
	NodeStatusLocked     NodeStatus = "locked"
	NodeStatusUnlocked   NodeStatus = "unlocked"
	NodeStatusInProgress NodeStatus = "in_progress"
	NodeStatusMastered   NodeStatus = "mastered"
	NodeStatusTestedOut  NodeStatus = "tested_out"
)

type UnlockRule struct {
	Type      string  `json:"type"` // "no_prerequisite", "mastery_threshold", "phase_complete", "quiz_pass"
	TopicID   string  `json:"topic_id,omitempty"`
	MinRating float64 `json:"min_rating,omitempty"`
	PhaseID   string  `json:"phase_id,omitempty"`
}

type RoadmapTemplate struct {
	ID         uuid.UUID     `json:"id"`
	TargetRole string        `json:"target_role"`
	Source     RoadmapSource `json:"source"`
	Version    int           `json:"version"`
	CreatedAt  time.Time     `json:"created_at"`
}

type RoadmapPhase struct {
	ID                uuid.UUID       `json:"id"`
	RoadmapTemplateID uuid.UUID       `json:"roadmap_template_id"`
	Sequence          int             `json:"sequence"`
	Title             string          `json:"title"`
	UnlockRule        json.RawMessage `json:"unlock_rule"`
	CreatedAt         time.Time       `json:"created_at"`
}

type RoadmapNode struct {
	ID               uuid.UUID       `json:"id"`
	PhaseID          uuid.UUID       `json:"phase_id"`
	TopicID          string          `json:"topic_id"`
	Sequence         int             `json:"sequence"`
	UnlockRule       json.RawMessage `json:"unlock_rule"`
	TutorialIDs      []uuid.UUID     `json:"tutorial_ids"`
	PracticeTopicIDs []string        `json:"practice_topic_ids"`
	CreatedAt        time.Time       `json:"created_at"`
}

type UserRoadmap struct {
	ID                uuid.UUID     `json:"id"`
	UserID            uuid.UUID     `json:"user_id"`
	RoadmapTemplateID uuid.UUID     `json:"roadmap_template_id"`
	Status            RoadmapStatus `json:"status"`
	CurrentPhaseID    *uuid.UUID    `json:"current_phase_id"`
	CreatedAt         time.Time     `json:"created_at"`
}

type UserRoadmapNodeProgress struct {
	ID            uuid.UUID  `json:"id"`
	UserRoadmapID uuid.UUID  `json:"user_roadmap_id"`
	NodeID        uuid.UUID  `json:"node_id"`
	Status        NodeStatus `json:"status"`
	UnlockedAt    *time.Time `json:"unlocked_at,omitempty"`
	MasteredAt    *time.Time `json:"mastered_at,omitempty"`
}

// Tutorial represents a learning resource (article, video, interactive lesson).
type Tutorial struct {
	ID               uuid.UUID `json:"id"`
	Source           string    `json:"source"`
	SourceURL        string    `json:"source_url"`
	Title            string    `json:"title"`
	TopicID          *string   `json:"topic_id,omitempty"`
	TopicTags        []string  `json:"topic_tags"`
	Type             string    `json:"type"`       // "article" | "video" | "interactive" | "playlist"
	Difficulty       string    `json:"difficulty"` // "beginner" | "intermediate" | "advanced"
	EstimatedMinutes int       `json:"estimated_minutes"`
	Summary          string    `json:"summary"`
	LicenseNote      string    `json:"license_note,omitempty"`
	ThumbnailURL     string    `json:"thumbnail_url,omitempty"`
	Status           string    `json:"status"` // "COMPLETED" | "UNREAD"
}

// RoadmapSubsection represents a single topic / node inside a roadmap section.
type RoadmapSubsection struct {
	NodeID       uuid.UUID        `json:"node_id"`
	TopicID      string           `json:"topic_id"`
	Title        string           `json:"title"`
	Sequence     int              `json:"sequence"`
	Status       NodeStatus       `json:"status"` // "locked", "unlocked", "in_progress", "mastered", "tested_out"
	UserRating   float64          `json:"user_rating"`
	TargetRating float64          `json:"target_rating"`
	MasteryScore float64          `json:"mastery_score"`
	Tutorials    []Tutorial       `json:"tutorials"`
	Questions    []ProblemPayload `json:"questions"`
}

// RoadmapSection represents a single progressive section (phase) visible to the user.
type RoadmapSection struct {
	PhaseID            uuid.UUID           `json:"phase_id"`
	Sequence           int                 `json:"sequence"`
	Title              string              `json:"title"`
	Status             string              `json:"status"` // "ACTIVE", "COMPLETED", "LOCKED"
	ProgressPercentage float64             `json:"progress_percentage"`
	Subsections        []RoadmapSubsection `json:"subsections"`
}

// UpcomingSectionPreview represents minimal metadata for locked upcoming sections.
type UpcomingSectionPreview struct {
	Sequence int    `json:"sequence"`
	Title    string `json:"title"`
	Status   string `json:"status"` // "LOCKED"
}

// RoadmapCurrentResponse is the primary dynamic roadmap payload fetched by the frontend.
// Only the current active section is populated with full tutorials and questions;
// subsequent sections are locked until the current section is mastered.
type RoadmapCurrentResponse struct {
	RoadmapID        uuid.UUID                `json:"roadmap_id"`
	UserID           string                   `json:"user_id"`
	UserRating       float64                  `json:"user_rating"`
	TargetRole       string                   `json:"target_role"`
	Status           RoadmapStatus            `json:"status"`
	TotalSections    int                      `json:"total_sections"`
	CurrentSection   *RoadmapSection          `json:"current_section"`
	UpcomingSections []UpcomingSectionPreview `json:"upcoming_sections"`
}

type RoadmapFull struct {
	Template RoadmapTemplate    `json:"template"`
	Phases   []RoadmapPhaseFull `json:"phases"`
}

type RoadmapPhaseFull struct {
	Phase RoadmapPhase      `json:"phase"`
	Nodes []RoadmapNodeFull `json:"nodes"`
}

type RoadmapNodeFull struct {
	Node     RoadmapNode              `json:"node"`
	Progress *UserRoadmapNodeProgress `json:"progress,omitempty"`
}

type CompleteNodeRequest struct {
	NodeID uuid.UUID `json:"node_id"`
}

type TestOutRequest struct {
	NodeID uuid.UUID `json:"node_id"`
}

File: frontend/src/mocks/fixtures/roadmap.ts
/**
 * A full 3-section roadmap mirroring `models.RoadmapCurrentResponse`
 * (backend/internal/models/roadmap.go): one ACTIVE section with fully
 * populated subsections (tutorials + questions), two LOCKED upcoming
 * section previews. Topic ids/names are pulled straight from the real
 * curriculum graph (`data/topics.json`) so this lines up with what the
 * live backend would actually return.
 */

import type {
  RoadmapCurrentResponse,
  RoadmapSubsection,
  Tutorial,
  UpcomingSectionPreview,
} from "@/lib/api/types";
import { PROBLEMS } from "./problems";

function questionsFor(topicId: string, count = 3) {
  const matches = PROBLEMS.filter((p) => p.topic === topicId);
  return matches.slice(0, count).map((p) => ({ ...p, statement: undefined, test_cases: undefined }));
}

function tutorial(
  id: string,
  topicId: string,
  title: string,
  opts: Partial<Tutorial> = {},
): Tutorial {
  return {
    id,
    source: opts.source ?? "neetcode",
    source_url: opts.source_url ?? `https://example.com/tutorials/${id}`,
    title,
    topic_id: topicId,
    topic_tags: opts.topic_tags ?? [topicId],
    type: opts.type ?? "article",
    difficulty: opts.difficulty ?? "beginner",
    estimated_minutes: opts.estimated_minutes ?? 12,
    summary: opts.summary ?? `A focused walkthrough of ${title.toLowerCase()}, with worked examples.`,
    license_note: opts.license_note,
    thumbnail_url: opts.thumbnail_url,
    status: opts.status ?? "UNREAD",
  };
}

function subsection(
  nodeId: string,
  topicId: string,
  title: string,
  sequence: number,
  status: RoadmapSubsection["status"],
  userRating: number,
  targetRating: number,
  masteryScore: number,
  tutorials: Tutorial[],
): RoadmapSubsection {
  return {
    node_id: nodeId,
    topic_id: topicId,
    title,
    sequence,
    status,
    user_rating: userRating,
    target_rating: targetRating,
    mastery_score: masteryScore,
    tutorials,
    questions: questionsFor(topicId),
  };
}

const ACTIVE_SUBSECTIONS: RoadmapSubsection[] = [
  subsection(
    "00000000-0000-4000-8000-000000000101",
    "foundations",
    "Foundations & Implementation",
    1,
    "mastered",
    1450,
    1200,
    1,
    [
      tutorial("t-101", "foundations", "Reading Constraints Like a Pro", {
        type: "article",
        estimated_minutes: 8,
        status: "COMPLETED",
      }),
      tutorial("t-102", "foundations", "Simulation Problems Walkthrough", {
        type: "video",
        estimated_minutes: 15,
        status: "COMPLETED",
      }),
    ],
  ),
  subsection(
    "00000000-0000-4000-8000-000000000102",
    "arrays-hashing",
    "Arrays & Hashing",
    2,
    "mastered",
    1480,
    1300,
    95,
    [
      tutorial("t-103", "arrays-hashing", "Hash Maps in 10 Minutes", {
        type: "video",
        estimated_minutes: 10,
        status: "COMPLETED",
      }),
    ],
  ),
  subsection(
    "00000000-0000-4000-8000-000000000103",
    "two-pointers",
    "Two Pointers",
    3,
    "in_progress",
    1310,
    1400,
    62,
    [
      tutorial("t-104", "two-pointers", "The Two-Pointer Pattern", {
        type: "article",
        estimated_minutes: 12,
        status: "COMPLETED",
      }),
      tutorial("t-105", "two-pointers", "Opposite-Direction Pointers, Visually", {
        type: "interactive",
        estimated_minutes: 20,
        status: "UNREAD",
      }),
    ],
  ),
  subsection(
    "00000000-0000-4000-8000-000000000104",
    "sliding-window",
    "Sliding Window",
    4,
    "unlocked",
    1180,
    1450,
    15,
    [
      tutorial("t-106", "sliding-window", "Fixed vs Variable Windows", {
        type: "article",
        estimated_minutes: 14,
        status: "UNREAD",
      }),
    ],
  ),
  subsection(
    "00000000-0000-4000-8000-000000000105",
    "stack-queues",
    "Stacks & Queues",
    5,
    "locked",
    0,
    1500,
    0,
    [
      tutorial("t-107", "stack-queues", "Monotonic Stacks From Scratch", {
        type: "article",
        estimated_minutes: 18,
        status: "UNREAD",
      }),
    ],
  ),
  subsection(
    "00000000-0000-4000-8000-000000000106",
    "binary-search",
    "Binary Search",
    6,
    "locked",
    0,
    1550,
    0,
    [
      tutorial("t-108", "binary-search", "Binary Search on the Answer", {
        type: "article",
        estimated_minutes: 16,
        status: "UNREAD",
      }),
    ],
  ),
];

const UPCOMING: UpcomingSectionPreview[] = [
  { sequence: 2, title: "Trees, Graphs & Traversal", status: "LOCKED" },
  { sequence: 3, title: "Dynamic Programming & Greedy", status: "LOCKED" },
];

export function buildRoadmap(): RoadmapCurrentResponse {
  return {
    roadmap_id: "00000000-0000-4000-8000-0000000000f1",
    user_id: "user-mock-001",
    user_rating: 1340,
    target_role: "Software Engineer - DSA & Problem Solving",
    status: "active",
    total_sections: 3,
    current_section: {
      phase_id: "00000000-0000-4000-8000-0000000000a1",
      sequence: 1,
      title: "Foundations & Linear Structures",
      status: "ACTIVE",
      progress_percentage: 42,
      subsections: ACTIVE_SUBSECTIONS,
    },
    upcoming_sections: UPCOMING,
  };
}

/** Mutable in-memory copy the handlers read/write against (node unlock progression, etc). */
export const roadmapState: { current: RoadmapCurrentResponse } = {
  current: buildRoadmap(),
};

export function findSubsection(nodeId: string): RoadmapSubsection | undefined {
  return roadmapState.current.current_section?.subsections.find((s) => s.node_id === nodeId);
}
